server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -453,6 +453,15 @@ public void applyClusterState(ClusterChangedEvent event) {
                 if (newMaster) {
                     finalizeSnapshotDeletionFromPreviousMaster(event.state());
                 }
+            } else if (snapshotCompletionListeners.isEmpty() == false) {
+                // We have snapshot listeners but are not the master any more. Fail all waiting listeners except for those that already
+                // have their snapshots finalizing (those that are already finalizing will fail on their own from to update the cluster
+                // state).
+                for (Snapshot snapshot : Set.copyOf(snapshotCompletionListeners.keySet())) {
```

Comment:
Gotta copy the keyset because `failSnapshotCompletionListeners` modifies the map